### ISSUE TITLE
automatic action string interpolation is not documented in Sphinx doc

### ISSUE THREAD

**Issue by vlcinsky** [pydoit/doit#193]
I was surprised to see, that the following code works:

```python
def task_compile_requirements():
    cmd = "pip-compile --output-file %(targets)s %(dependencies)s"
    return {
        "file_dep": ["requirements.in"],
        "actions": [cmd],
        "targets": ["requirements.txt"],
    }
```

I did not know, that the action string `cmd` is automatically
interpolated by:
- `targets`
- `dependencies`
- `changed`

Such an information exists for `CmdAction` in code, but I did not find such information in Sphinx documentation.

I would propose adding it and explain that CmdAction and action string are automatically interpolated by values of `targets`, `dependencies` and `changed`.

**Comment by schettino72** [comment]
I surprised this is not explicitly mentioned in the docs although the very first example of CmdAction uses this http://pydoit.org/tasks.html#cmd-action.

Pull-requests improving docs are welcome.

**Comment by vlcinsky** [comment]
@shettino72 yes, you are right. The doc explains it sufficiently.

I am (really) found of reading documentation properly.
But in this case it happened before 2015 and current quick search for related keywords using Sphinx search feature failed to locate the description. And I failed being too lazy to reread the doc more carefully.

Sorry. I learnt, that even daily use of the tool does not make me an expert.

Experimenting with the options I came to following task:

```python
from doit.action import CmdAction


def task_cmdstring_interpolation():
    """Show few cases where command string interpolation takes place."""

    def build_5():
        cmd = "echo 3:inter D: %(dependencies)s T: %(targets)s CH: %(changed)s"
        return cmd

    def build_6(dependencies, targets, changed):
        cmd = "echo 4:inter D: %(dependencies)s T: %(targets)s CH: %(changed)s"
        dct = {
            "d": str(dependencies),
            "t": str(targets),
            "ch": str(changed),
        }
        return cmd + " " + str(dct)

    return {
        "file_dep": ["one.in", "two.in"],
        "targets": ["t1.out", "t2.out", "t3.out"],
        "actions": [
            "echo 1:KISS",
            "echo 2:inter D: %(dependencies)s T: %(targets)s CH: %(changed)s",
            CmdAction("echo 3:KISS"),
            CmdAction("echo 4:inter %(dependencies)s %(targets)s %(changed)s"),
            CmdAction(build_5),
            CmdAction(build_6)
        ],
        "verbosity": 2,
    }
```

1. simle command string. Must not conflict with string formatting using `%`.
2. use of implicit string interpolation.
3. The same applies to use of `CmdAction` when providing command string.
4. `CmdAction` also gets command string interpolated.
5. `CmdAction` can get the command string provided by function call, the interpolation then applies to returned string.
6. The function `build_6` is called with `dependencies`, `targets` and `changed` arguments. Interpolation will be applied on returned string anyway.

If the doc deserves any update at all (I am not sure about it), I would go this way:

- if command is provided in form of a string (in contrast to list of command arguments), it always interpolates the `dependencies`, `targets` and `changed`.
- for that reason, the command string must be properly formatted not to clash with python sprintf style string formatting.
- `dependencies`, `targets` and `changed` values for interpolation are strings created from lists by concatenated with " "


I am not sure, if these modification shall appear in the doc as it is already explaining it rather well and my version could add too many details and harm readibility.

Another takeaway for me from this issue is, it will be worth rereading the doc again, as I found there many jewels like `getargs`, `pathlib` etc. It is excellent piece of work. Thanks for that.

**Comment by schettino72** [comment]
I am reopening this because I also searched for keywords and could not find references to it. 
So there is definitely something to be improved.

> I am not sure, if these modification shall appear in the doc as it is already explaining it rather well and my version could add too many details and harm readibility.

I originally wrote the docs to serve both as guide/tutorial and reference.
But after adding many features this approach is not good enough any more because it is impossible to balance full readability with a complete reference.

I would like to re-write the docs with a different approach but that needs time (that i am not really willing to dedicated right now).

**Comment by schettino72** [comment]
A nice talk I watched about writing docs: https://www.youtube.com/watch?v=t4vKPhjcMZg

**Comment by vlcinsky** [comment]
@schettino72 Nice video.

Quick start is a type of doc I  would add, it is sort of pre-tutorial and shall make very first impression of "yes, this is what I need, lets dive into it".

I have a dream, all python projects will stop using `make` and will use `doit` instead. Because Python and portability counts.

For this reason I have a proposal: Let's update the docs together. I would take care of Quick start and Tutorials, because that creates new users.

You would take care of Reference, because nobody has better knowledge of the facts inside of doit. 

The rest (How-tos and Discussion stuff) could appear as needed.

I am also rather busy and such task could last few month or even a year, but with clear vision we can make a plan and finally execute.

Think of it.

**Comment by schettino72** [comment]
Cool. Glad you want to contribute.

I guess for reference the docs are already pretty good. Maybe add some docs on internals.

There is an issue #123 about writing a tutorial. I have some ideas on how it should be done.. if you are interested we can discuss about that.

**Comment by schettino72** [comment]
fixed on #196
